README.md
# treasury_direct

[![Build Status](https://travis-ci.com/larryaasen/treasury_direct.svg?branch=master)](https://travis-ci.com/larryaasen/treasury_direct)    [![codecov](https://codecov.io/gh/larryaasen/treasury_direct/branch/master/graph/badge.svg)](https://codecov.io/gh/larryaasen/treasury_direct) [![pub package](https://img.shields.io/pub/v/treasury_direct.svg)](https://pub.dartlang.org/packages/treasury_direct)

A Dart library package to access the debt of the United States from the US Treasury site. This package can be used in a command line app or in a Flutter app.

## Example

```dart
import 'package:treasury_direct/treasury_direct.dart';
import 'package:treasury_direct/src/debt_entry.dart';

void main() async {
  final td = TreasuryDirect();
  final list = await td.downloadDebtFeedAsync(pagesize: 15);

  print('Total possible rows: ${list.totalRows}');

  print('Total row received: ${list.mostRecentList.length}');
  for (DebtEntry entry in list.mostRecentList) {
    print('${DebtEntry.dateFormatted(entry.effectiveDate)}: ${DebtEntry.currencyShortened(entry.totalDebt, false)}');
  }
}
```

Expected output:
```text
Total possible rows: 7156
Total row received: 7156
Oct 7, 2021: $28.43T
Oct 6, 2021: $28.43T
Oct 5, 2021: $28.43T
Oct 4, 2021: $28.43T
Oct 1, 2021: $28.43T
Sep 30, 2021: $28.43T
Sep 29, 2021: $28.43T
Sep 28, 2021: $28.43T
Sep 27, 2021: $28.43T
Sep 24, 2021: $28.43T
Sep 23, 2021: $28.43T
Sep 22, 2021: $28.43T
Sep 21, 2021: $28.43T
Sep 20, 2021: $28.43T
Sep 17, 2021: $28.43T
...
```

## Debt to the Penny API Docs:
https://fiscaldata.treasury.gov/datasets/debt-to-the-penny/


CHANGELOG.md
## 3.1.1

* Updated dependency version for http.

## 3.1.0

* Updated dependency package versions, and allowed Dart version 3.

## 3.0.2

* Minor updates to resolve static analysis issues.

## 3.0.1

* Minor updates to resolve static analysis issues. Updated example.

## 3.0.0

* Upgraded to null safety.

## 2.4.0

* Changed API domain from www.transparency.treasury.gov to api.fiscaldata.treasury.gov.

## 2.3.0

* Fixed bug when intragov_hold_amt was null.

## 2.2.0

* Fixed crash when intragov_hold_amt was null.

## 2.1.0

* Changed package http version requirement to >=0.12.2 and updated documentation.

## 2.0.1

* Downgraded package async to 2.4.0 because flutter_driver still requires version 2.4.1.

## 2.0.0

* Updated the Treasury API endpoint to use the new endpoint: https://www.transparency.treasury.gov/services/api/fiscal_service/v2/accounting/od/debt_to_penny. Documentation here: https://fiscaldata.treasury.gov/datasets/debt-to-the-penny/.

## 0.1.4

* Minor updates based on Health suggestions from pub.dev, utilizing pedanctic.

## 0.1.3

* Changed async version from 2.4.0 to 2.3.0 to support the current version of Flutter.

## 0.1.2

* Minor update

## 0.1.1
*  Added exported files that were missing from export.

## 0.1.0

* Initial Open Source release.


test/treasury_direct_test.dart
import 'package:test/test.dart';

import 'package:treasury_direct/treasury_direct.dart';

/// Tests the Treasury Direct library.
void main() {
  test('download 10', () async {
    final td = TreasuryDirect();

    final list = await td.downloadDebtFeedAsync();
    expect(list.totalRows, greaterThan(3700));
    expect(list.mostRecentList!.length, equals(10));
    for (var entry in list.mostRecentList!) {
      validateDebtEntry(entry);
    }

    final entry1 = list.mostRecentList![0];
    final entry2 = list.mostRecentList![1];
    final entry3 = list.mostRecentList![2];
    expect(entry1.date()!.isAtSameMomentAs(entry2.date()!), isFalse);
    expect(entry1.date()!.isAfter(entry2.date()!), isTrue);
    expect(entry2.date()!.isAfter(entry3.date()!), isTrue);
  });

  test('download 1024', () async {
    final td = TreasuryDirect();

    final list = await td.downloadDebtFeedAsync(pagesize: 1024);
    expect(list.totalRows, greaterThan(3700));
    expect(list.mostRecentList!.length, equals(1024));
  });

  test('urlString', () async {
    final td = TreasuryDirect();

    expect(td.urlString(),
        'https://api.fiscaldata.treasury.gov/services/api/fiscal_service/v2/accounting/od/debt_to_penny?page[number]=1&page[size]=10&sort=-record_date');
    expect(td.urlString(pagesize: 25),
        'https://api.fiscaldata.treasury.gov/services/api/fiscal_service/v2/accounting/od/debt_to_penny?page[number]=1&page[size]=25&sort=-record_date');
    expect(td.urlString(pagenum: 25),
        'https://api.fiscaldata.treasury.gov/services/api/fiscal_service/v2/accounting/od/debt_to_penny?page[number]=25&page[size]=10&sort=-record_date');
    expect(td.urlString(pagesize: 25, pagenum: 25),
        'https://api.fiscaldata.treasury.gov/services/api/fiscal_service/v2/accounting/od/debt_to_penny?page[number]=25&page[size]=25&sort=-record_date');
  });

  test('DebtEntry.shortened', () {
    expect(DebtEntry.currencyShortened(-10.2, true), equals('-\$10.20'));
    expect(DebtEntry.currencyShortened(-10.12, true), equals('-\$10.12'));
    expect(DebtEntry.currencyShortened(-0.1, true), equals('-10¢'));
    expect(DebtEntry.currencyShortened(0.0, true), equals('0¢'));
    expect(DebtEntry.currencyShortened(0.10, true), equals('+10¢'));
    expect(DebtEntry.currencyShortened(1.11, true), equals('+\$1.11'));
    expect(DebtEntry.currencyShortened(12.12, true), equals('+\$12.12'));
    expect(DebtEntry.currencyShortened(123.13, true), equals('+\$123.13'));
    expect(DebtEntry.currencyShortened(1234.14, true), equals('+\$1.23K'));
    expect(DebtEntry.currencyShortened(12345.15, true), equals('+\$12.35K'));
    expect(DebtEntry.currencyShortened(123456.16, true), equals('+\$123.46K'));
    expect(DebtEntry.currencyShortened(1234567.17, true), equals('+\$1.23M'));
    expect(DebtEntry.currencyShortened(12345678.18, true), equals('+\$12.35M'));
    expect(
        DebtEntry.currencyShortened(123456789.19, true), equals('+\$123.46M'));
    expect(
        DebtEntry.currencyShortened(1234567890.20, true), equals('+\$1.23B'));
    expect(
        DebtEntry.currencyShortened(12345678901.21, true), equals('+\$12.35B'));
    expect(DebtEntry.currencyShortened(123456789012.22, true),
        equals('+\$123.46B'));
    expect(DebtEntry.currencyShortened(1234567890123.23, true),
        equals('+\$1.23T'));
  });
}

void validateDebtEntry(DebtEntry entry) {
  expect(entry.effectiveDate!.length, greaterThanOrEqualTo(10));
  expect(entry.governmentHoldings, greaterThan(10000));
  expect(entry.totalDebt, greaterThan(10000));
  expect(entry.change, isNot(0));
}


lib/src/debt_list.dart
import 'debt_entry.dart';

/// A list of debt entries.
class DebtList {
  final int? totalRows;
  final List<DebtEntry>? mostRecentList;

  DebtList({this.totalRows = 0, this.mostRecentList});

  /// Create a list from JSON data.
  static DebtList listFromJSON(Map<String, dynamic> json) {
    final meta = json['meta'];
    final totalCount = meta['total-count'] as int?;

    final data = json['data'];

    final newEntries = <DebtEntry>[];

    if (data is List<dynamic>) {
      for (final entry in data) {
        if (entry is Map<String, dynamic>) {
          final newEntry = DebtEntry.debtFromJSON(entry);
          newEntries.add(newEntry);
        }
      }
    }

    // Compute the diffs
    var index = 0;
    for (final entryLater in newEntries) {
      if (index + 1 < newEntries.length) {
        final entryEarlier = newEntries[index + 1];
        entryLater.change = entryLater.totalDebt! - entryEarlier.totalDebt!;
      }
      index++;
    }

    final newList = DebtList(totalRows: totalCount, mostRecentList: newEntries);

    return newList;
  }
}


lib/src/debt_entry.dart
import 'package:intl/intl.dart';

/// One entry for the debt for the day.
/// Reference: https://fiscaldata.treasury.gov/datasets/debt-to-the-penny/debt-to-the-penny
class DebtEntry {
  /// The date that data was published.
  String? effectiveDate;

  ///  Holdings	Government Account Series (GAS) securities held by Government trust funds, revolving funds, and special funds; and Federal Financing Bank (FFB) securities.
  double? governmentHoldings;

  ///Total of intragovernmental holdings and debt held by the public.
  double? totalDebt;

  /// The amount of change from the previous entry.
  double change;

  DebtEntry({
    this.effectiveDate = '',
    this.governmentHoldings = 0.0,
    this.totalDebt = 0.0,
    this.change = 0.0,
  });

  static final DateFormat _fmt = DateFormat('y-M-d');

  // Keep a cached version of the effectiveDate
  DateTime? _date;

  /// The effectiveDate as a [DateTime].
  DateTime? date() {
    _date ??= convertToDate(effectiveDate);
    return _date;
  }

  /// A string representation of this object.
  @override
  String toString() => 'DebtEntry{totalDebt: $totalDebt}';

  /// Convert map to [DebtEntry].
  static DebtEntry debtFromJSON(Map<String, dynamic> json) {
    String effectiveDate = '';
    try {
      effectiveDate = json['record_date'] as String;
    } on Exception catch (e) {
      print('DebtEntry.debtFromJSON: record_date exception: $e');
      print('DebtEntry.debtFromJSON: json: $json');
    }

    double governmentHoldings = 0.0, totalDebt = 0.0;
    try {
      final amt = json['intragov_hold_amt'] as String?;
      if (amt != null && amt != 'null') {
        governmentHoldings = double.parse(amt);
      }
    } catch (e) {
      print('DebtEntry.debtFromJSON: intragov_hold_amt exception: $e');
      print('DebtEntry.debtFromJSON: json: $json');
    }

    try {
      final amt = json['tot_pub_debt_out_amt'] as String?;
      if (amt != null && amt != 'null') {
        totalDebt = double.parse(amt);
      }
    } catch (e) {
      print('DebtEntry.debtFromJSON: tot_pub_debt_out_amt exception: $e');
      print('DebtEntry.debtFromJSON: json: $json');
    }

    final change = 1.0;

    return DebtEntry(
      effectiveDate: effectiveDate,
      governmentHoldings: governmentHoldings,
      totalDebt: totalDebt,
      change: change,
    );
  }

  /// Convert value to date.
  static DateTime? convertToDate(String? sDate) {
    if (sDate == null) return null;

    DateTime newDate;
    try {
      newDate = _fmt.parse(sDate);
    } catch (e) {
      print('DebtEntry: cannot parse date ($sDate): $e');
      return null;
    }

    return newDate;
  }

  /// Format the date.
  static String? dateFormatted(String value) {
    final date = convertToDate(value);
    if (date == null) return null;

    final dateString = DateFormat('MMM d, y').format(date);
    return dateString;
  }

  /// Convert current value to a shortened version.
  static String currencyShortened(double value, bool includePrefix) {
    final short = _shortened(value, value);

    if (includePrefix) {
      final prefix = value > 0.0
          ? '+'
          : value < 0.0
              ? '-'
              : '';
      return prefix + short;
    }

    return short;
  }

  /// Convert value to trillions.
  static double trillions(double value) {
    return value / 1000000000000;
  }

  static String _shortened(double value, double originalValue) {
    if (value < 0.0) {
      return _shortened(-value, originalValue);
    }
    if (value < 1) {
      return '${(value * 100).toInt()}¢';
    }
    if (value < 1000) {
      return '\$${value.toStringAsFixed(2)}';
    }
    if (value < 1000000) {
      return '\$${(value / 1000).toStringAsFixed(2)}K';
    }

    if (value < 1000000000) {
      return '\$${(value / 1000000).toStringAsFixed(2)}M';
    }
    if (value < 1000000000000) {
      return '\$${(value / 1000000000).toStringAsFixed(2)}B';
    }
    if (value < 1000000000000000) {
      return '\$${(value / 1000000000000).toStringAsFixed(2)}T';
    }

    return value.toStringAsFixed(2);
  }
}

/*

{
	"data": [{
			"record_date": "2020-07-16",
			"debt_held_public_amt": "20611921079550.83",
			"intragov_hold_amt": "5917339138678.45",
			"tot_pub_debt_out_amt": "26529260218229.28",
			"src_line_nbr": "1",
			"record_fiscal_year": "2020",
			"record_fiscal_quarter": "4",
			"record_calendar_year": "2020",
			"record_calendar_quarter": "3",
			"record_calendar_month": "07",
			"record_calendar_day": "16"
		},
		{
			"record_date": "2020-07-15",
			"debt_held_public_amt": "20574380830507.80",
			"intragov_hold_amt": "5911091609553.12",
			"tot_pub_debt_out_amt": "26485472440060.92",
			"src_line_nbr": "1",
			"record_fiscal_year": "2020",
			"record_fiscal_quarter": "4",
			"record_calendar_year": "2020",
			"record_calendar_quarter": "3",
			"record_calendar_month": "07",
			"record_calendar_day": "15"
		}
	],
	"meta": {
		"count": 2,
		"labels": {
			"record_date": "Record Date",
			"debt_held_public_amt": "Debt Held by the Public",
			"intragov_hold_amt": "Intragovernmental Holdings",
			"tot_pub_debt_out_amt": "Total Public Debt Outstanding",
			"src_line_nbr": "Source Line Number",
			"record_fiscal_year": "Fiscal Year",
			"record_fiscal_quarter": "Fiscal Quarter Number",
			"record_calendar_year": "Calendar Year",
			"record_calendar_quarter": "Calendar Quarter Number",
			"record_calendar_month": "Calendar Month Number",
			"record_calendar_day": "Calendar Day Number"
		},
		"dataTypes": {
			"record_date": "DATE",
			"debt_held_public_amt": "CURRENCY",
			"intragov_hold_amt": "CURRENCY",
			"tot_pub_debt_out_amt": "CURRENCY",
			"src_line_nbr": "NUMBER",
			"record_fiscal_year": "YEAR",
			"record_fiscal_quarter": "QUARTER",
			"record_calendar_year": "YEAR",
			"record_calendar_quarter": "QUARTER",
			"record_calendar_month": "MONTH",
			"record_calendar_day": "DAY"
		},
		"dataFormats": {
			"record_date": "YYYY-MM-DD",
			"debt_held_public_amt": "$10.20",
			"intragov_hold_amt": "$10.20",
			"tot_pub_debt_out_amt": "$10.20",
			"src_line_nbr": "10.2",
			"record_fiscal_year": "YYYY",
			"record_fiscal_quarter": "Q",
			"record_calendar_year": "YYYY",
			"record_calendar_quarter": "Q",
			"record_calendar_month": "MM",
			"record_calendar_day": "DD"
		},
		"total-count": 3716,
		"total-pages": 1858
	},
	"links": {
		"self": "&page%5Bnumber%5D=1&page%5Bsize%5D=2",
		"first": "&page%5Bnumber%5D=1&page%5Bsize%5D=2",
		"prev": null,
		"next": "&page%5Bnumber%5D=2&page%5Bsize%5D=2",
		"last": "&page%5Bnumber%5D=1858&page%5Bsize%5D=2"
	}
}

 */
